utils/bitfield/src/unvalidated.rs

```diff
@@ -0,0 +1,78 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use super::{BitField, Result};
+use encoding::serde_bytes;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+
+/// A trait for types that can produce a `&BitField` (or fail to do so).
+/// Generalizes over `&BitField` and `&mut UnvalidatedBitField`.
+pub trait Validate<'a> {
+    fn validate(self) -> Result<&'a BitField>;
+}
+
+impl<'a> Validate<'a> for &'a mut UnvalidatedBitField {
+    /// Validates the RLE+ encoding of the bit field, returning a shared
+    /// reference to the decoded bit field.
+    fn validate(self) -> Result<&'a BitField> {
+        self.validate_mut().map(|bf| &*bf)
+    }
+}
+
+impl<'a> Validate<'a> for &'a BitField {
+    fn validate(self) -> Result<&'a BitField> {
+        Ok(self)
+    }
+}
+
+/// A bit field that may not yet have been validated for valid RLE+.
+/// Used to defer this validation step until when the bit field is
+/// first used, rather than at deserialization.
+#[derive(Debug)]
+pub enum UnvalidatedBitField {
+    Validated(BitField),
+    Unvalidated(Vec<u8>),
+}
+
+impl UnvalidatedBitField {
+    /// Validates the RLE+ encoding of the bit field, returning a unique
+    /// reference to the decoded bit field.
+    pub fn validate_mut(&mut self) -> Result<&mut BitField> {
+        if let Self::Unvalidated(bytes) = self {
+            *self = Self::Validated(BitField::from_bytes(&bytes)?);
+        }
+
+        match self {
+            Self::Validated(bf) => Ok(bf),
+            Self::Unvalidated(_) => unreachable!(),
+        }
+    }
+}
+
+impl From<BitField> for UnvalidatedBitField {
+    fn from(bf: BitField) -> Self {
+        Self::Validated(bf)
+    }
+}
+
+impl Serialize for UnvalidatedBitField {
+    fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        match self {
+            Self::Validated(bf) => serde_bytes::serialize(&bf.to_bytes(), serializer),
+            Self::Unvalidated(bytes) => serde_bytes::serialize(&bytes, serializer),
+        }
+    }
+}
```

Comment:
optionally, can remove impl for this with serde `untagged` https://serde.rs/enum-representations.html#untagged (prob best to keep the deserialize either way)